File: core/parser/xray.py
"""
Парсер конфигураций формата Xray (JSON схема с REALITY, Vision и Mux).
"""

import json
from typing import Any, Dict, List

from core.config.models import ProfileConfig, ProtocolType
from core.parser.base import BaseParser, ParseResult
from utils.logger import log_call


class XrayParser(BaseParser):
    """Парсер для Xray JSON конфигураций."""

    format_name: str = "xray"

    PROTOCOL_MAP = {
        "vless": ProtocolType.VLESS,
        "vmess": ProtocolType.VMESS,
        "trojan": ProtocolType.TROJAN,
        "shadowsocks": ProtocolType.SHADOWSOCKS,
        "wireguard": ProtocolType.WIREGUARD,
        "freedom": ProtocolType.DIRECT,
        "blackhole": ProtocolType.BLOCK,
    }

    @log_call
    def parse(self, content: str) -> ParseResult:
        """
        Разбирает JSON-конфигурацию Xray.

        Args:
            content: Текст JSON Xray.

        Returns:
            ParseResult: Результат парсинга.
        """
        result = ParseResult(format_name=self.format_name)

        try:
            data = json.loads(content)
        except json.JSONDecodeError as exc:
            result.errors.append(f"Ошибка синтаксиса Xray JSON: {exc}")
            return result

        if not isinstance(data, dict):
            result.errors.append("Корневой элемент Xray JSON должен быть словарем.")
            return result

        result.raw_config = data
        outbounds = data.get("outbounds", [])

        if not isinstance(outbounds, list):
            result.errors.append("Секция 'outbounds' Xray должна быть списком.")
            return result

        for item in outbounds:
            if not isinstance(item, dict):
                continue

            protocol_str = item.get("protocol", "").lower()
            tag = item.get("tag", "xray-outbound")

            if protocol_str in ("freedom", "blackhole", "dns"):
                continue

            protocol = self.PROTOCOL_MAP.get(protocol_str, ProtocolType.VLESS)
            settings = item.get("settings", {})
            stream_settings = item.get("streamSettings", {})

            # Вынимаем хост и порт
            server = "127.0.0.1"
            port = 443

            vnext = settings.get("vnext", [])
            servers = settings.get("servers", [])
            if vnext and isinstance(vnext, list):
                server = vnext[0].get("address", server)
                port = int(vnext[0].get("port", port))
            elif servers and isinstance(servers, list):
                server = servers[0].get("address", server)
                port = int(servers[0].get("port", port))

            profile = ProfileConfig(
                name=tag,
                protocol=protocol,
                server=server,
                server_port=port,
                settings={
                    "xray_settings": settings,
                    "stream_settings": stream_settings,
                    "mux": item.get("mux", {}),
                },
                tag=tag,
            )
            result.profiles.append(profile)

        return result

File: core_manager/src/parser/xray.rs
//! Парсер Xray JSON конфигураций на Rust.

use crate::models::{ProfileConfig, ProtocolType};
use crate::parser::{ConfigParser, ParseResult};

pub struct XrayParser;

impl ConfigParser for XrayParser {
    fn format_name(&self) -> &'static str {
        "xray"
    }

    fn parse(&self, content: &str) -> ParseResult {
        let mut result = ParseResult::new(self.format_name());

        let json_val: serde_json::Value = match serde_json::from_str(content) {
            Ok(v) => v,
            Err(err) => {
                result.errors.push(format!("Ошибка синтаксиса Xray JSON: {}", err));
                return result;
            }
        };

        result.raw_config = Some(json_val.clone());

        if let Some(outbounds) = json_val.get("outbounds").and_then(|o| o.as_array()) {
            for item in outbounds {
                let proto_str = item.get("protocol").and_then(|p| p.as_str()).unwrap_or("vless").to_lowercase();
                let tag = item.get("tag").and_then(|t| t.as_str()).unwrap_or("xray-proxy");

                if ["freedom", "blackhole", "dns"].contains(&proto_str.as_str()) {
                    continue;
                }

                let settings = item.get("settings").cloned().unwrap_or_default();
                let stream = item.get("streamSettings").cloned().unwrap_or_default();

                let mut server = "127.0.0.1".to_string();
                let mut port: u16 = 443;

                if let Some(vnext) = settings.get("vnext").and_then(|v| v.as_array()).and_then(|arr| arr.first()) {
                    if let Some(addr) = vnext.get("address").and_then(|a| a.as_str()) {
                        server = addr.to_string();
                    }
                    if let Some(p) = vnext.get("port").and_then(|p| p.as_u64()) {
                        port = p as u16;
                    }
                }

                let protocol = match proto_str.as_str() {
                    "vless" => ProtocolType::Vless,
                    "vmess" => ProtocolType::Vmess,
                    "trojan" => ProtocolType::Trojan,
                    "shadowsocks" => ProtocolType::Shadowsocks,
                    "wireguard" => ProtocolType::Wireguard,
                    _ => ProtocolType::Vless,
                };

                let mut p = ProfileConfig::new(tag, protocol, server, port);
                p.settings = serde_json::json!({
                    "xray_settings": settings,
                    "stream_settings": stream,
                    "mux": item.get("mux")
                });
                result.profiles.push(p);
            }
        }

        result
    }
}
